src/tests/Loader/classloader/RefFields/Validate.cs

```diff
@@ -19,6 +19,25 @@ public static void Validate_Invalid_RefField_Fails()
         Assert.Throws<TypeLoadException>(() => { var t = typeof(InvalidObjectRefRefFieldOverlap); });
     }
 
+    [Fact]
+    public static void Validate_ConservativeGC_Types()
+    {
+        // These types represent scenarios where a conservative GC is needed.
+        // Since Mono always runs with a conservative GC, these types are permissable.
```

Comment:
Not that I mind getting something for free, but is this a behavior difference we want to allow?

The field layout check in mono has bits to spare for native pointer and byref fields if we cared to do the check.
https://github.com/dotnet/runtime/blob/348e06c4c2a0c4758e4fb02cfee27ecbedcb1a87/src/mono/mono/metadata/class-init.c#L1944

Reply:
That is possible. I was going to update Mono, but realized since it has a conservative GC this support is technically okay. More than willing to block it though if the Mono team decides it should be done.

Reply:
I will say the complexity here is non-trivial because we need to know more than just if it contains a byref. We need to know where that byref field exists to check for invalid overlap.

Reply:
@AaronRobinsonMSFT 
The algorithm in mono is to (working recursively - flattening any valuetypes we see along the way):
1. create a byte array the size of the toplevel struct type and fill it with zeroes.   
2. for each field, if it's a reference type write a 0x01 on each byte that the field occupies; otherwise write 0x02.  When writing if we see a non-zero byte at that offset already, and it's different from what we're writing, throw a TLE.   

For example:
```csharp
struct X {
  [FieldOffset(0)] object o ; //   writes 1 1 1 1 1 1 1 1 starting at index 0 in the byte array on x64
  [FieldOffset(3)] byte b; // wants to write 2 at index 3 in the byte array.  But there's already a 1 there - TLE
```

So I think we can do a similar thing for byref, unless I'm missing something:
```csharp
  [FieldOffset(0)] ref int x; // writes 3 3 3 3 3 3 3 3 starting at offset 0
  [FieldOffset(5)] short int y; // wants to write 2 2 at index 5 ===> TLE
```

(I'm assuming `object o` and `ref int y` aren't allowed to overlap, either)

Reply:
> (I'm assuming object o and ref int y aren't allowed to overlap, either)

@lambdageek Yep. That has been validated.

I've applied the fix to mono. I am much less confident in how I fixed this. The previous fixes in mono _seemed_ simpler so would appreciate a look here.